File: basicImplementationDotNet/Controllers/TokenController.cs
using BasicImplementationDotNet.Models;
using BasicImplementationDotNet.Services;
using Microsoft.AspNetCore.Mvc;

namespace BasicImplementationDotNet.Controllers;

[ApiController]
[Route("[controller]")]
public class TokenController : ControllerBase
{
    private readonly ILogger _logger;
    private readonly IMonoovaApiService _monoovaApiService;

    public TokenController(
        ILogger<TokenController> logger,
        IMonoovaApiService monoovaApiService)
    {
        _logger = logger;
        _monoovaApiService = monoovaApiService;
    }

    /// <summary>
    /// use this endpoint to create a client session token with sample input from PopulateRequestSample() method
    /// </summary>
    [HttpPost]
    public async Task<ActionResult<CreateClientSessionTokenResponse>> GetClientSessionTokenAsync(
        CancellationToken cancellationToken)
    {
        var request = PopulateRequestSample();
        _logger.LogInformation("Received request to create Client Session Token for {request}", request);
        var loginResponse = await _monoovaApiService.LoginToMonoovaAsync(cancellationToken);
        if (loginResponse.Errors != null)
            return BadRequest(loginResponse);

        _logger.LogInformation("Successfully logged in with bearer token: {token}", loginResponse.Token);
        
        var tokenResponse = await _monoovaApiService.CreateClientSessionTokenAsync(loginResponse.Token, request, cancellationToken);
        if (tokenResponse.Errors != null)
            return BadRequest(tokenResponse);

        _logger.LogInformation("Successfully created Client Session Token: {token}", tokenResponse.ClientToken);
        return tokenResponse;
    }

    
    /// <summary>
    /// use this endpoint to create a client session token with custom input
    /// </summary>
    [HttpPost]
    [Route("TokenCustomInput")]
    public async Task<ActionResult<CreateClientSessionTokenResponse>> CreateClientSessionTokenWithCustomPayloadAsync(
        [FromBody] CreateClientSessionTokenRequest request,
        CancellationToken cancellationToken)
    {
        _logger.LogInformation("Received request to create Client Session Token for {request}", request);
        var loginResponse = await _monoovaApiService.LoginToMonoovaAsync(cancellationToken);
        if (loginResponse.Errors != null)
            return BadRequest(loginResponse);
        _logger.LogInformation("Successfully logged in with bearer token: {token}", loginResponse.Token);
        
        var tokenResponse = await _monoovaApiService.CreateClientSessionTokenAsync(loginResponse.Token, request, cancellationToken);
        if (tokenResponse.Errors != null)
            return BadRequest(tokenResponse);

        _logger.LogInformation("Successfully created Client Session Token: {token}", tokenResponse.ClientToken);
        return tokenResponse;
    }


    private static CreateClientSessionTokenRequest PopulateRequestSample()
    {
        var request = new CreateClientSessionTokenRequest
        {
            Customer = new Customer
            {
                CustomerId = "testCustomerId",
                EmailAddress = "test@gmail.com",
                BillingAddress = new BillingAddress
                {
                    FirstName = "testFirstName",
                    LastName = "testLastName",
                    PostalCode = "2000",
                    Street = new List<string> { "100 Sussex street" },
                    CountryCode = "AU",
                    Suburb = "SYDNEY",
                    State = "NSW"
                }
            },
            PaymentDetails = new PaymentDetails
            {
                CardType = "Visa",
                Description = "sample token create",
                SaveOnSuccess = false
            },
            Amount = new Amount
            {
                CurrencyAmount = "1.00"
            },
            ClientTransactionUniqueReference = Guid.NewGuid().ToString()
        };
        return request;
    }
}

File: basicImplementation/src/controllers/tokenController.ts
// tokenController.ts
import { Request, Response } from 'express'
import { v4 as uuidv4 } from 'uuid'
import * as monoovaService from '../services/monoovaService'
import { MonoovaLoginResponse } from '../models/MonoovaLoginResponse'
import { CreateClientSessionTokenRequest } from '../models/CreateClientSessionTokenRequest'

/**
 * @openapi
 * /token:
 *   post:
 *     summary: Create a new token with a predefined input
 *     tags:
 *       - Token
 *     requestBody:
 *       required: false
 *     responses:
 *       200:
 *         description: Successfully created token
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/CreateClientSessionTokenResponse'
 *       400:
 *         description: Bad Request
 *       500:
 *         description: Internal Server Error
 */
export const createToken = async (req: Request, res: Response) => {
  try {
    const createClientSessionTokenRequest: CreateClientSessionTokenRequest = {
      customer: {
        customerId: 'testCustomerId',
        emailAddress: 'test@gmail.com',
        billingAddress: {
          firstName: 'testFirstName',
          lastName: 'testLastName',
          postalCode: '2000',
          street: ['100 Sussex street'],
          countryCode: 'AU',
          suburb: 'SYDNEY',
          state: 'NSW'
        }
      },
      paymentDetails: {
        cardType: 'Visa',
        description: 'sample token create',
        saveOnSuccess: false
      },
      amount: {
        currencyAmount: '1.00'
      },
      clientTransactionUniqueReference: uuidv4()
    }

    const loginResponse: MonoovaLoginResponse = await monoovaService.loginToMonoova()

    const createTokenResponse = await monoovaService.createClientSessionToken(
      loginResponse.token,
      createClientSessionTokenRequest
    )

    if (createTokenResponse.errors && createTokenResponse.errors.length > 0) {
      const errorDetails = createTokenResponse.errors.map((error) => error.errorMessage).join('; ')
      return res.status(400).json({ error: errorDetails })
    }

    res.json(createTokenResponse)
  } catch (error) {
    console.error('Error processing /token endpoint:', error)
    if (error instanceof Error) {
      res.status(500).json({ error: error.message })
    } else {
      res.status(500).json({ error: 'Internal Server Error' })
    }
  }
}

/**
 * @openapi
 * /createTokenCustomInput:
 *   post:
 *     summary: Create a new token
 *     tags:
 *       - Token
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/CreateClientSessionTokenRequest'
 *     responses:
 *       200:
 *         description: Successfully created token
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/CreateClientSessionTokenResponse'
 *       400:
 *         description: Bad Request
 *       500:
 *         description: Internal Server Error
 */
export const createTokenCustomInput = async (req: Request, res: Response) => {
  try {
    const loginResponse: MonoovaLoginResponse = await monoovaService.loginToMonoova()
    const createClientSessionTokenRequest: CreateClientSessionTokenRequest = req.body
    const createTokenResponse = await monoovaService.createClientSessionToken(
      loginResponse.token,
      createClientSessionTokenRequest
    )

    if (createTokenResponse.errors && createTokenResponse.errors.length > 0) {
      return res.status(400).json(createTokenResponse)
    }

    res.json(createTokenResponse)
  } catch (error) {
    console.error('Error processing /token endpoint:', error)
    if (error instanceof Error) {
      res.status(500).json({ error: error.message })
    } else {
      res.status(500).json({ error: 'Internal Server Error' })
    }
  }
}
